### PR TITLE
Extract Hitbox functionality to dedicated class

### PR DESCRIPTION
**Pull request by NimuKaT**
* Implement base Hitbox class
* Implement rectangular Hitbox class
* Implement circular Hitbox class
* Implement rectangle to rectangle hitbox detection
* Implement circle to circle hitbox detection
* Implemente circle to rectangel hitbox detection
* Update Wall to use Hitbox class
* Change Wall to use collision detection with relative movement
* Change Game to use relative collision detection
* Change Character to use relative collision detection
* Add absolute position hitbox detection for Entites
* Change variable names in game.ts

### CODE REVIEW

File: src/server/character.ts [hunk 1]
```diff
@@ -45,17 +45,19 @@ export class Character extends Entity {
         const g = getGame();
         let movex = Math.round(this._sx * delta);
         let movey = Math.round(this._sy * delta);
-        while (g.collision(this._x + movex, this._y, this._r) && movex != 0) {
+        while (g.collision(this._hitbox, movex, 0) && movex != 0) {
             this._sx = 0;
             movex -= Math.sign(movex);
         }
         this._x += movex;
+        this._updateHitbox();
```
Comment (Naidah) [inline src/server/character.ts:53]: Add move function to make movement and hitbox update unit function

File: src/server/character.ts [hunk 2]
```diff
@@ -45,17 +45,19 @@ export class Character extends Entity {
...
 
-        while (g.collision(this._x, this._y + movey, this._r) && movey != 0) {
+        while (g.collision(this._hitbox, 0, movey) && movey != 0) {
             this._sy = 0;
             movey -= Math.sign(movey);
         }
         this._y += movey;
+        this._updateHitbox();
```
Comment (Naidah) [inline src/server/character.ts:60]: As above

File: src/server/entity.ts [hunk 3]
```diff
@@ -22,6 +26,11 @@ export class Entity {
   update(delta: number): void {
     this._x += this._sx * delta;
     this._y += this._sy * delta;
+    this._updateHitbox();
```
Comment (Naidah) [inline src/server/entity.ts:29]: Unary move function, as above

File: src/server/hitbox.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        this._y = y;
+    }
+    // Relative position update
+    move(dx: number, dy: number) {
+        this._x += dx;
+        this._y += dy;
+    }
+
+    // Test if the passed hitbox will collide if it moves by the given velocity
+    testMovement (hitbox: Hitbox, dx: number, dy: number) : boolean {
```
Comment (Naidah) [inline src/server/hitbox.ts:30]: default value of 0, 0 for dx and dy (will help projectiles, etc)

File: src/server/hitbox.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        hitbox.move(dx, dy);
+        let result = this.hasCollision(hitbox);
+        hitbox.move(-dx, -dy);
+        return result;
+    }
+
+    // Checks if a collisions occurs
+    hasCollision (hitbox: Hitbox) : boolean {return false}; // Should be an abstract method
```
Comment (Naidah) [inline src/server/hitbox.ts]:
```suggestion
    abstract hasCollision (hitbox: Hitbox) : boolean;
```

File: src/server/hitbox.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    // Selects the collision function based on hitbox type
+    hasCollision (hitbox: Hitbox) : boolean {
+        if (hitbox instanceof RectHitbox) {
+            return this.rectOnRect(hitbox);
+        }
+        else {
+            return this.circleOnRect(hitbox as CircleHitbox);
+        }
+    }
```
Comment (Naidah) [inline src/server/hitbox.ts:61]: Possible to use method overloading?

File: src/server/hitbox.ts (hunk 5 above)
Comment (NimuKaT) [reply src/server/hitbox.ts]: @Naidah I was thinking of doing this but it seems typescript doesn't like abstract methods in non-abstract class. Let me know if I'm wrong.

**Pull request merged**